### ISSUE TITLE
MobileServiceSyncTable method PurgeAsync with Force flag

### ISSUE THREAD

**Issue by ltaljaard** [Azure/azure-mobile-apps-net-client#109]
I need to purge entries on the client-side (mobile device) where there is a pending operation. To do that I need to call a version of PurgeAsync that accepts the "force" flag. There is no version available that takes a LINQ Where clause (IMobileServiceTableQuery) as parameter. The only one available for my purposes is one that takes a normal SQL string as the query/selection parameter. It seems like a step backwards to me to have to start writing my LINQ queries as plain-text now. Please consider one of the following as a solution to this problem:

Option 1: Expose a new version of PurgeAsync that takes an IMobileServiceTableQuery as well as the force flag as parameters

Option 2: Expose the following internal method of class MobileServiceTableQueryProvider to the outside world so that one can call it to prepare a plain-text string version of your query from your Linq query - this is already being used internally in the mobile sync client-side code so there is no reason that it could not be exposed quickly and easily: the method that will need to be exposed is: "internal string ToODataString<T>(IMobileServiceTableQuery<T> query)"

PS: This is how the above ToODataString() method is currently already being used in internal class MobileServiceSyncTable<T>. If you go for option 2 above then this is what I would like to have access to from my own code i.e. from outside of the Mobile Services library:

```
public Task PurgeAsync<U>(string queryId, IMobileServiceTableQuery<U> query, CancellationToken cancellationToken)
        {
            if (query == null)
            {
                throw new ArgumentNullException("query");
            }
            string queryString = this.queryProvider.ToODataString(query);

            return this.PurgeAsync(queryId, queryString, false, cancellationToken);
        }
```

**Comment by adrianhall** [comment]
Can you give us a use case outside of the code.  What are you actually trying to achieve with this functionality?  (I understand the functionality - I want to understand the reasoning behind it)

**Comment by ltaljaard** [comment]
I simply want to delete a record/object that has pending operations on it. There are currently six versions of the PugeAsync() method available for public use, of which two take the force flag. So my reasons are the same as the use cases for which the force flag is currently already available (i.e. nothing new). But the only problem with the existing versions that already have the force flag is that the first of the two existing versions wipes all the data in the table so that won't work for me, and the second of the two existing versions only takes a query expressed as a string. Hence my request.

**Comment by ltaljaard** [comment]
Hi Adrian, I've made some changes to my local copy of this github project and it seems that it's working (with some changes to class PurgeAction method HandleDirtyTable because currently it only supports deleting ALL operations against a local table and it's ignoring the filter i.e./e.g. the object ID that's being passed in - amongst some other small changes). But before I submit it for your review to be merged let me give you my exact use case.

Step 1: I have some custom logic that pulls some data (which represents let's say an Agency server-side object) into memory of the local device from the server before giving the user the choice whether they want to save it locally or not. I get the server data with a PullAsync call to the server. Then, if the user decides to save the data locally, I use UpdateAsync() to save the data locally. Why do I do that? Because, firstly, using InsertAsync() locally with the data in memory would cause a new instance of the object to be created on the server, which is not the intention. And secondly because I don't want to make another PullAsync() call over the network just to save this data locally, seeing that it is already present in my local memory - so such a resource intensive call seems completely unnecessary.

Step 2: The above approach that I'm currently using i.e. using UpdateAsync() instead of PullAsync() seems to work fine in other aspects, except for when the user decides to delete the object from the local store before a PushAsync() has occurred. In that case, when PurgeAsync() is called, an exception is thrown because there is still a pending operation on the object that I inserted in the local store in step 1 above. That's why I need a Force flag on the PurgeAsync() method (the force flag would allow me to delete the local object despite the fact that there is a pending "update" operation for it).

So my primary question at this point is this: does the rationale that I set out in step 1 above hold up? Is it acceptable to call UpdateAsync() in this way or is there a fundamental reason why PullAsync() i.e. a seemingly unnecessary second round-trip to the server is required to safely store the object locally? (if the only problem with this approach is the local purging part that I described in part 2 above then I can submit my changes to my local project to get this working for your review).
